torch/_higher_order_ops/cond.py

```diff
@@ -376,6 +381,33 @@ def cond_op_dense(pred, true_fn, false_fn, operands):
         return false_fn(*operands)
 
 
+# WAR for https://github.com/pytorch/pytorch/issues/140322
+@cond_op.py_impl(CUDAGraphCaptureControlFlowOpDispatchMode)
+def cond_op_cudagraph(mode, pred, true_fn, false_fn, operands):
+    assert torch.cuda.is_available() and torch.cuda.is_current_stream_capturing()
+    if not mode.inside_already_warmed_up_op:
+        for fn in (true_fn, false_fn):
+            if fn not in mode.warmed_up_control_flow_ops:
+                warmup_mode = ControlFlowOpWarmupDispatchMode()
+                with warmup_mode:
+                    fn(*operands)
+                mode.warmed_up_control_flow_ops.add(fn)
+    mode.inside_already_warmed_up_op = True
+    # Re-enter this mode in order to handle nested control flow
+    # operations
+    with mode:
+        output = if_else_node(pred, true_fn, false_fn, operands)
+    mode.inside_already_warmed_up_op = False
+    return output
+
+
+# WAR for https://github.com/pytorch/pytorch/issues/140322
+@cond_op.py_impl(ControlFlowOpWarmupDispatchMode)
+def cond_op_warmup(mode, pred, true_fn, false_fn, operands):
+    assert torch.cuda.is_available() and torch.cuda.is_current_stream_capturing()
```

Comment:
Basically, I need to warm up both sides of an if-else statement (as well as the body of a while loop, which won't run during warmup if the condition function returns false). However, we don't know a-priori whether or not all ops will be "warmed up". Running the while loop body when cond_fn is false may result in the program entering an invalid state. Similarly, running both side of an if-else statement can also result in a program entering an invalid state. This is because torch.while_loop() and torch.cond() can modify external tensors.

The *only* way in the CUDA API to "stub out" a cuda kernel, such that it does not actually execute (and therefore does not cause any side effects that may cause the program to enter an invalid state) is to use relaxed stream capture to a cuda graph (that cuda graph is then discarded). Meanwhile, the warmup "unsafe actions" like creating a cublasHandle, cudnnHandle, etc. will be done correctly.

Basically, I want to make sure that, e.g., if someone warms up only the left side of a torch.cond() function, that stream capture will still succeed on the right hand side. This is sort of captured in the following test case:

https://github.com/pytorch/pytorch/pull/140979/files#diff-ce3a6ac7310a0a222e113af989463fe8f5bd7394a13a9cbfa58cfc0a7ceb494eR54-R72

(This test case fails to do any warm up at all because I was ambitious, but if you add `f(pred, x)` before `with torch.cuda.graph()`, it is clear to see that eager mode execution will fail to warm up the STFT operation, which does an unsafe cuda api call via cuModuleLoadData() when falling back from fake tensors to real tensors to compute the output shape correctly.) Let me quote your comment here about "I don't think we should do first run support of cuda graph": https://github.com/pytorch/pytorch/pull/140979#discussion_r1874063316 I would tend to agree with you on that. It's just that, there is no guarantee that all ops will be warmed up in the presence of data-dependent control flow right now.

The messiness due to this is by far the most annoying thing about this PR. If we had an interface that took a python function that the user wanted to do stream capture on, that would work pretty well, since then I could basically call the function once in eager mode (*not* using stream capture), with a special TorchDispatchMode that could do relaxed stream capture on torch.cond() and torch.while_loop()'s functions just for the sake of warming them up, and *then* do cuda stream capture for real the second time around. The current interface, where torch.cuda.graph() is a context manager that has no direct access to the code that runs between __enter__ and __exit__, I am kind of boned for warming things up. Maybe we could use a new interface where users have to pass a function, I'm not sure.

Reply:
What do you think about initially requiring the user do:

``` 
with ControlFlowOpWarmupDispatchMode():
      fn(x)
with torch.cuda.graph(g):
     fn(x)
```

I know the extra api call is maybe a bit annoying, but this would be handled automatically with torch.compile invocation. Additionally, anyone using both torch control flow operators and cudagraph is going to be a bit more on the power side. 

thoughts ?

Reply:
This is a great idea. I've implemented, but not pushed it.

However, the torch.compile() cache won't be hit if there is a difference in the current TorchDispatchMode stack, apparently since PR https://github.com/pytorch/pytorch/pull/131828 (but I'm not 100% certain). Since the control flow was done with ControlFlowOpWarmupDispatchMode(), but the actual stream capture does not use that dispatch mode, this ends up not working, haha.

I'm digging deeper into torch dynamo internals to try to figure out the right way to go about this...